pagefaultgames/pokerogue
# [Feature] Add Prestige Mode (compiled version)

> [!NOTE]
> This PR refers to https://github.com/pagefaultgames/pokerogue/issues/1545
> I got kindly asked to compile the PRs into a single one. I'm sorry for the inconvenience.
> You can take a look at this issue https://github.com/pagefaultgames/pokerogue/issues/1545 if you want the full context, but this PR should be enough, as I'm trying to make it self-contained for future reference.

cc:
- @molker
- @Xavion3 since this might be of interest to you

## What are the changes?

This PR allows to introduce the **Prestiges** to the game.

The idea of the **Prestiges** is to allow the player to start over the _Classic_ mode with some increased difficulties. The game will include some modifiers to make the game harder. It contains multiple levels of difficulty, and each level will be harder than the previous one. You need to complete the game in the current difficulty to unlock the next one.

To unlock **Prestiges**, you first need to complete the game in the _Classic_ mode. Once you've done that, you will be able to start over the game in the _Prestige_ mode.

Currently, the game will includes 8 levels of **Prestiges**, referring to the 8 gyms of a Pokemon game.

## Why am I doing these changes?




I found a [very interesting ticket talking](https://github.com/pagefaultgames/pokerogue/issues/1283) about this kind of system, which I believe makes a lot of sense in roguelite games. We can find similar systems in games like [_Slay The Spire_](https://slay-the-spire.fandom.com/wiki/Ascension) for instance.

Also, I really like the idea of adding difficulty and challenges to the game, without going for the endless mode. I believe this can bring a lot of replayability to the game, and I'm sure some players will enjoy this kind of challenge.

I also like this idea because we can easily extend it with more levels, or mode modifiers if needed, which means more difficulties/challenges in the future.

### But why would someone want to play in Prestige mode?

Short answer: for the PRESTIGE!

On another note, I believe it makes it quite exciting to start over the game with some increased difficulties.
I personally prefer games with challenges, than games that are using the length of the game as a way to make it harder. Of course, this is personal, but I believe some players will enjoy this kind of challenge.

Also, winning the Classic mode in a given prestige allows to unlock an achievement, and an associated Egg voucher.

It can be a good way to challenge yourself and see how far you can go in the game. What is nice about this system is that we can potentially extend it with more levels, or mode modifiers if needed, which means more difficulties/challenges.

## What did change?



This PR introduces the **Prestige** system to the game. Among these changes, we can find the following:
- UI
  - Allows to unlock the Prestige mode whenever the player completes the game in the Classic mode for the first time
  - Allows to unlock the next Prestige level whenever the player completes the game in the current Prestige level
  - Avoids unlocking the next Prestige level if the player is at the last Prestige level
  - Adds a Prestige level selection menu whenever the player chooses the Classic mode, hidden whenever Prestige mode is not unlocked
  - Adds the current Prestige level in the UI when playing the game
  - Adds a tooltip to show the current modifiers when hovering the Prestige level in the UI when playing the game
  - Indicates the Prestige level in the save slot
  - Indicates the maximum Prestige level completed in the Game stats
- Modifiers:
  - Decreases the starter party points whenever the player starts a new game in Prestige mode
  - Increases enemy pokemons stats whenever the player starts a new game in Prestige mode
  - Increases the shop item prices whenever the player starts a new game in Prestige mode
  - Decreases the party luck whenever the player starts a new game in Prestige mode
  - Decreases the experience gain whenever the player starts a new game in Prestige mode
- Rewards
  - Adds achievement for each prestige
  - Adds associated egg voucher for each prestige completed

Note: I didn't dive into level design, so we might want to tweak the values of the modifiers.

### Screenshots/Videos



- Completing the Classic mode for the first time and unlocking Prestige mode

  https://github.com/pagefaultgames/pokerogue/assets/57403591/e553b436-6458-48c3-9afd-105239cfb52b

- Completing the Classic mode in a given prestige mode and unlocking the next level

  https://github.com/pagefaultgames/pokerogue/assets/57403591/ae6bffa8-188c-439f-9c0c-352fd11fdfb5

## How to test the changes?




You can finish the game first to unlock the Prestige mode. You can do so by giving yourself a good pokemon with good move starting at wave 199 against Eternatus. Here are the changes I made to `sec/overrides.ts`:
```typescript
// ...
export const STARTING_WAVE_OVERRIDE: integer = 199;
// ...
export const STARTING_LEVEL_OVERRIDE: integer = 50000;
// ...
export const STARTER_SPECIES_OVERRIDE: Species | integer = Species.MEWTWO;
// ...
export const MOVESET_OVERRIDE: Array<Moves> = [Moves.PSYCHIC, Moves.FLAMETHROWER, Moves.THUNDERBOLT, Moves.ICE_BEAM];
// ...
export const OPP_SPECIES_OVERRIDE: Species | integer = Species.ETERNATUS;
// ...
export const OPP_MOVESET_OVERRIDE: Array<Moves> = [Moves.SPLASH, Moves.SPLASH, Moves.SPLASH, Moves.SPLASH];
// ...
export const STARTING_HELD_ITEMS_OVERRIDE: Array<ModifierOverride> = [
  { name: "MULTI_LENS", count: 3 }
];
```

Once you have it, you can actually start playing with Prestiges. You can do so by going to the Classic mode, and you will see a new screen asking you to choose the Prestige level you want to play with. You can try the different features explained above.

You can also manually modify the values of the modifiers under `src/system/prestige.ts`.

## Checklist

What do you think?
